### PR TITLE
Improve handling empty segments in urls according to RFC3986

### LINKED ISSUE #926: __truediv__ to append a path drops final /
### Describe the bug

We have an internal api that cares about whether the url ends in a '/'. 
This may be a poor design, but such things exist. 

When a path is added to a url with the / operator, the final '/' is stripped.

```
import yarl
from yarl import URL
url = URL(f'http://example.com') / 'foo/bar/baz/'
print(url)
```

### To Reproduce

```
from yarl import URL
url = URL(f'http://example.com') / 'foo/bar/baz/'
assert str(url).endswith('/'), "where did my slash go?"

url = URL(f'http://example.com') / 'foo' / 'bar'/ 'baz' / ''
assert str(url).endswith('/'), "still no slash?"
```

### Expected behavior

The final slash should be preserved when appending a path with the __truediv__ operator. 
This would be in accord with the Principle of Least Surprise. 

### Logs/tracebacks

```python-traceback
>>> from yarl import URL
>>> url = URL(f'http://example.com') / 'foo/bar/baz/'
>>> assert str(url).endswith('/'), "where did my slash go?"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: where did my slash go?
>>> 
>>> url = URL(f'http://example.com') / 'foo' / 'bar'/ 'baz' / ''
>>> assert str(url).endswith('/'), "still no slash?"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: still no slash?
```

### Python Version

```console
$ python --version
Python 3.11.4
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.4
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/matthew/venvs/wxp/lib/python3.11/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /home/matthew/venvs/wxp/lib/python3.11/site-packages
Requires: idna, multidict
Required-by: aio-pika, aiohttp, aiormq
```

### OS

$ uname -a
Linux mattswxp 5.15.0-76-generic #83-Ubuntu SMP Thu Jun 15 19:16:32 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### LINKED ISSUE #984: Regression: Trailing slash for yarl.URL ("http://localhost") / "path" / "" is dropped in 1.9.4
### Describe the bug

yarl 1.8.2 allowed to construct URLs with trailing slashes like that:
```
>>> str(yarl.URL ("http://localhost") / "path" / "") 
'http://localhost/path/'
```

yarl 1.9.4 drops the slash from that path. Possibly related to https://github.com/aio-libs/yarl/issues/862 :
```
>>> str(yarl.URL ("http://localhost") / "path" / "") 
'http://localhost/path'
```

### To Reproduce

Try 1.8.2 vs 1.9.4 with the code above.

### Expected behavior

Preserve the slash

### Logs/tracebacks

```python-traceback
-
```

### Python Version

```console
Python 3.11.6
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.4
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /home/alex/.local/share/virtualenvs/widget.backend-6HfWXuQ_/lib/python3.11/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

Ubuntu 23.10

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by commonism**
## What do these changes do?

The changes honor empty segments when joining urls
e.g.
```python
URL("https://web.archive.org/web/") / "https://github.com/"
```

Picking up the PR #1023 to get this merged.

##  Are there changes in behavior for the user?

The changes to URL.join and URL.joinpath do not strip empty segments any longer.

## Related issue number

Fixes #984
Fixes #926

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes

### CODE REVIEW

**Comment by commonism** [comment]
@Dreamsorcerer can you help me figure out the issue with the codecov/ checks?

I can see it ran the unit tests, but failed to create proper coverage files?
e.g. https://app.codecov.io/gh/aio-libs/yarl/pull/1026/blob/tests/test_url.py#L859

**Comment by Dreamsorcerer** [comment]
> @Dreamsorcerer can you help me figure out the issue with the codecov/ checks?
> 
> I can see it ran the unit tests, but failed to create proper coverage files? e.g. https://app.codecov.io/gh/aio-libs/yarl/pull/1026/blob/tests/test_url.py#L859

I'd change the fail option to true, then it'll be obvious if any uploads have failed (and which jobs need to be rerun):
https://github.com/aio-libs/yarl/blob/master/.github/workflows/ci-cd.yml#L340

File: yarl/_url.py [hunk 1]
```diff
@@ -713,28 +713,36 @@ def _validate_authority_uri_abs_path(host, path):
...
-                parsed.append(seg)
+            path = path if encoded else self._PATH_QUOTER(path)
+            if not (
+                segments := [
+                    segment for segment in reversed(path.split("/")) if segment != "."
+                ]
+            ):
+                continue
+            # remove trailing empty segment for all but the last path
+            parsed += segments[1:] if not last and segments[0] == "" else segments
```
Comment (Dreamsorcerer) [inline yarl/_url.py:737]: `last` is always `0`?

Comment (Dreamsorcerer) [reply yarl/_url.py:737]: Nevermind

File: yarl/_url.py [hunk 2]
```diff
@@ -713,28 +713,36 @@ def _validate_authority_uri_abs_path(host, path):
...
-            if "/" in seg:
-                parsed += (
-                    sub for sub in reversed(seg.split("/")) if sub and sub != "."
+                    f"Appending path {path!r} starting from slash is forbidden"
                 )
-            elif seg != ".":
-                parsed.append(seg)
+            path = path if encoded else self._PATH_QUOTER(path)
+            if not (
+                segments := [
```
Comment (Dreamsorcerer) [inline yarl/_url.py:731]: I think this would be cleaner to put the assignment before the if, rather than having it wrap over so many lines and have that much complexity in one statement.

**Comment by commonism** [comment]
> I'd change the fail option to true, then it'll be obvious if any uploads have failed (and which jobs need to be rerun): https://github.com/aio-libs/yarl/blob/master/.github/workflows/ci-cd.yml#L340

It's not an upload failing.
[Test (3.12, Y, ubuntu-latest, false)](https://github.com/aio-libs/yarl/actions/runs/9598259348/job/26469466683?pr=1026#logs)
All tests ran, upload succeeded.

Yet the only tests which are run evaluated are the tests which have a annotation to return None
[def test_authority_full()](https://app.codecov.io/gh/aio-libs/yarl/pull/1026?flags%5B0%5D=Py-3.12.4&src=pr&el=tree&filepath=tests%2Ftest_url.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr%20comments&utm_term=aio-libs#8b270584bb5bef11f7777857af051674-L201)

You can see test_url.py and tests/test_url.py exist.
https://app.codecov.io/gh/aio-libs/yarl/pull/1026/tree

Same issue: https://app.codecov.io/gh/aio-libs/yarl/pull/1025/tree
Same issue: https://app.codecov.io/gh/aio-libs/yarl/pull/1023/tree
Same issue: https://app.codecov.io/gh/aio-libs/yarl/pull/1019/tree
Okay: https://app.codecov.io/gh/aio-libs/yarl/pull/1018/tree
Okay: https://app.codecov.io/gh/aio-libs/yarl/pull/1016/tree

Same problem: https://app.codecov.io/gh/aio-libs/yarl/pull/1015/blob/tests/test_url.py

**Comment by Dreamsorcerer** [comment]
> Yet the only tests which are run evaluated are the tests which have a annotation to return None

Oh, that's because webknjaz added mypy coverage to the results. I think that's just confusing...

If you'd like to improve the situation, then you can add tests/ to coverage in pytest.ini and look at enabling the options we use elsewhere in .mypy.ini while fixing up the errors it highlights (maybe add a couple of options in each PR). You can use aiohttp-jinja2 as a reference: https://github.com/aio-libs/aiohttp-jinja2

**Comment by commonism** [comment]
> > Yet the only tests which are run evaluated are the tests which have a annotation to return None
> 
> Oh, that's because webknjaz added mypy coverage to the results. I think that's just confusing...

MyPy has it's own flag, the flag I chose [here](https://app.codecov.io/gh/aio-libs/yarl/pull/1026?flags%5B0%5D=Py-3.12.4&src=pr&el=tree&filepath=tests%2Ftest_url.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr%20comments&utm_term=aio-libs#8b270584bb5bef11f7777857af051674-L201) is Py-3.12.4, still showing the mypy results?
For [Base](https://app.codecov.io/gh/aio-libs/yarl/commit/7bca6cf1ba3673018b9065c6102f791a0a7516c0/tree?flags%5B0%5D=Py-3.12.4) the file is [missing](https://app.codecov.io/gh/aio-libs/yarl/commit/7bca6cf1ba3673018b9065c6102f791a0a7516c0/blob/test_url.py?flags%5B0%5D=Py-3.12.4).

May be better to have somebody look into this who has an idea how this is supposed to work.
Currently it's broken for every PR.

**Comment by Dreamsorcerer** [comment]
Yeah, but it's clearly mypy coverage. See above if you want to improve either coverage.

**Comment by commonism** [comment]
To my understanding …

The coverage reported by [mypy](https://app.codecov.io/gh/aio-libs/yarl/pull/1026/tree?flags%5B0%5D=MyPy) is "[tests/test_url.py](https://app.codecov.io/gh/aio-libs/yarl/pull/1026/blob/tests/test_url.py?flags%5B0%5D=MyPy)", the coverage reported by [python 3.12.4](https://app.codecov.io/gh/aio-libs/yarl/pull/1026/tree?search=&flags%5B0%5D=Py-3.12.4) is for "[test_url.py](https://app.codecov.io/gh/aio-libs/yarl/pull/1026/blob/test_url.py)".

The reported codecov numbers are combined values for unit testing and type checking.
The low value report is due to the mypy/typing use and does not reflect unit testing code coverage.

Due to a large number of single file posts, [codecov may rate limit](https://github.com/aio-libs/yarl/actions/runs/9640464656/job/26584495389#step:14:71) (seen in the scheduled CI runs), I guess this could be prevented by limiting the number of requests, combining all reports in a single upload.

**Comment by Dreamsorcerer** [comment]
Right, but if I understood correctly from the configs, tests/ also haven't been added to the unit test coverage report at all?

But, again, I'd enable the fail on error setting, so we don't have any missing uploads.

**Comment by Dreamsorcerer** [comment]
Oh, wait, I see what you mean. Codecov wouldn't load anything last time I looked. So, it looks like changing from codecov v3 to v4 resulted in the file paths changing for some reason...

**Comment by Dreamsorcerer** [comment]
Well, copying the config from aiohttp-jinja2, as I suggested, resolves the path, but now the yarl/ coverage seems to have disappeared: #1028.

**Comment by webknjaz** [comment]
@Dreamsorcerer there's a few points to check:
- what gets into codecov — "uploads" at https://app.codecov.io/gh/aio-libs/yarl/commit/7bca6cf1ba3673018b9065c6102f791a0a7516c0/tree?flags%5B0%5D=Py-3.12.4 has https://api.codecov.io/upload/gh/aio-libs/yarl/download?path=v4/raw/2024-06-14/F2143C32CD85E0BC827E552A5E512211/7bca6cf1ba3673018b9065c6102f791a0a7516c0/3eded8ad-f926-4cb5-b6a6-5cbce8a67443/48384b50-8891-45cb-8a42-87d70c33fd3e.txt which reveals what the Codecov CLI uploader sends to their server — it does only contain base filenames in the XML part, which is surprising
- then, I'd check what coverage.py produces in `coverage.xml` (it should be the same that it's in the payload
- if it's the fault of the coverage library and not Codecov, I'd try older versions — we've been attempting to update it recently in Ansible and hit some regressions related to forked subprocesses, for example

It is a good idea to bisect Codecov @ https://app.codecov.io/gh/aio-libs/yarl/commits.

Looks like https://app.codecov.io/gh/aio-libs/yarl/commit/60737a8bf57625cffb09674b788ab2f64320a448 is the last commit with proper paths and the following https://app.codecov.io/gh/aio-libs/yarl/commit/e74c1599d47e21523d3e78ad78966d9630b12cc0 is broken.

So it looks like https://github.com/aio-libs/yarl/pull/1007 might've influenced it, which could only come from the MyPy bump there or unpinned deps.

Comparing the last working (https://api.codecov.io/upload/gh/aio-libs/yarl/download?path=v4/raw/2024-06-06/F2143C32CD85E0BC827E552A5E512211/60737a8bf57625cffb09674b788ab2f64320a448/2bd0dbd5-af59-44bd-8e2b-479f8d54ef00.txt) and the first broken (https://api.codecov.io/upload/gh/aio-libs/yarl/download?path=v4/raw/2024-06-06/F2143C32CD85E0BC827E552A5E512211/e74c1599d47e21523d3e78ad78966d9630b12cc0/6a695c6c-3879-48c0-bf3d-62e2d036817a/d9f3f16e-284b-43ce-a191-ce045b11982f.txt) uploads (both from Py-3.11.9 @ ubuntu-latest), the first thing I noticed is that the file list before the XML chunk that was present earlier, disappeared. The second thing is that the older working upload had a `# path=./coverage.xml` comment and the newer one had `# path=/home/runner/work/yarl/yarl/coverage.xml`. Both have `<coverage version="7.5.3"` which means that the `coverage.py` version hasn't changed in between.

This points to some env changes, which might be versions of pytest/pytest-cov/codecov-action/codecov-cli etc.

**Comment by webknjaz** [comment]
@Dreamsorcerer you could also try producing an html report locally to see if `coverage.py` behaves well stand-alone.

**Comment by webknjaz** [comment]
Oh... There's actually more commits in between 60737a8bf57625cffb09674b788ab2f64320a448 and e74c1599d47e21523d3e78ad78966d9630b12cc0: https://github.com/aio-libs/yarl/commits/e74c1599d47e21523d3e78ad78966d9630b12cc0/.

And among those, there's a series of action bumps, which probably didn't have a chance to make uploads, being cancelled by newer merges starting more CI jobs. I compared our two commits' logs and the confirmed my suspicion:
```diff
- Download action repository 'codecov/codecov-action@v3.1.4' (SHA:eaaf4bedf32dbdc6b720b63067d99c4d77d6047d)
+ Download action repository 'codecov/codecov-action@v4' (SHA:125fc84a9a348dbcf27191600683ec096ec9021c)
```
*(https://github.com/aio-libs/yarl/actions/runs/9401318235/job/25893141828#step:1:41 vs. https://github.com/aio-libs/yarl/actions/runs/9405561693/job/25907242226#step:1:41)*

With that in mind, my new suspect is... :drum: :long_drum: :drum: ... https://github.com/aio-libs/yarl/pull/988!

**Comment by webknjaz** [comment]
Good news is that if you're persistent enough, it's possible to get them to merge your bugfix: https://github.com/codecov/uploader/pulls?q=sort%3Aupdated-desc+is%3Apr+author%3Awebknjaz+is%3Amerged / https://github.com/codecov/codecov-cli/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Amerged+author%3Awebknjaz.

**Comment by webknjaz** [comment]
To conclude, the current workaround for viewing the coverage would be `make cov; python -Im webbrowser htmlcov/index.html`.

**Comment by commonism** [comment]
Tried v3.1.4 - https://github.com/aio-libs/yarl/actions/runs/9691011881/job/26741939301
Can't say anything about the results as uploading fails due to rate limiting - https://github.com/codecov/codecov-action/issues/1293 may be related.

**Comment by webknjaz** [comment]
> Can't say anything about the results as uploading fails due to rate limiting - [codecov/codecov-action#1293](https://github.com/codecov/codecov-action/issues/1293) may be related.

I ended up hardcoding the token in pytest the other day (https://github.com/pytest-dev/pytest/pull/12516), to fight this problem. I think, we should do it everywhere for now.

**Review by Dreamsorcerer — APPROVED** [review]

File: CHANGES/1026.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Joining URLs with empty segments has been changed
+to match RFC3986 -- by :user:`commonism` and :user:`youtux`
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:2]: Sphinx has a dedicated role for RFCs.
```suggestion
to match :rfs:`3986` -- by :user:`commonism` and :user:`youtux`.
```

Comment (webknjaz) [reply CHANGES/1026.bugfix.rst:2]: Also, it would be a good idea to include more high-level context for the end users, that is not a reference to elsewhere but is more specific to what the change influences.

File: yarl/_url.py [hunk 4]
```diff
@@ -713,28 +713,35 @@ def _validate_authority_uri_abs_path(host, path):
...
         parsed.reverse()
-        old_path = self._val.path
-        if old_path:
-            parsed = [*old_path.rstrip("/").split("/"), *parsed]
+
+        if self._val.path:
+            # remove trailing empty segment
+            if (old := self._val.path.split("/"))[-1] == "":
+                del old[-1]
+            parsed = [*old, *parsed]
```
Comment (webknjaz) [inline yarl/_url.py:743]: This shouldn't need a code comment but better structure:
```suggestion
        if self._val.path and (old_path_segments := self._val.path.split("/")):
            old_path_cutoff = -1 if old_path_segments[-1] == "" else None
            parsed = [*old_path_segments[:old_path_cutoff], *parsed]
```

File: tests/test_url.py [hunk 5]
```diff
@@ -827,6 +839,40 @@ def test_joinpath(base, to_join, expected):
...
+        pytest.param("path", "a//", "path/a//", id="default_trailing-empty-segments"),
+        pytest.param("path", "a//b", "path/a//b", id="default_embedded-empty-segment"),
+    ],
+)
+def test_joinpath_empty_segments(base, to_join, expected):
+    url = URL(f"http://example.com/{base}")
+    assert (
+        str(url.joinpath(to_join))
+        == str(url / to_join)
+        == f"http://example.com/{expected}"
```
Comment (webknjaz) [inline tests/test_url.py:864]: Do we really want a double equality check here? It's always confusing how these are evaluated.

File: .github/workflows/ci-cd.yml [hunk 6]
```diff
@@ -339,7 +339,7 @@ jobs:
           OS-${{ runner.os }},
           VM-${{ matrix.os }},
           Py-${{ steps.python-install.outputs.python-version }}
-        fail_ci_if_error: false
+        fail_ci_if_error: true
```
Comment (webknjaz) [inline .github/workflows/ci-cd.yml:342]: Could you submit this separately?

File: tests/test_url.py (hunk 5 above)
Comment (Dreamsorcerer) [reply tests/test_url.py:864]: I don't mind how it's formatted, but I asked for the test to check that .joinpath() and the `/` operator are both producing the correct output.

Comment (webknjaz) [reply tests/test_url.py:864]: This comment wasn't about formatting but chained `==`. They often confuse me as I can never remember that corner case when they behave in unexpected ways.

**Comment by webknjaz** [comment]
Could you move the Codecov action update to a separate PR?
Also, the change notes are RST so it would be nice to use that markup. The RTD preview might be broken if you don't. It's a blocker, FYI.

**Comment by commonism** [comment]
Updated accordingly.
Where is the RTD preview located?

**Comment by webknjaz** [comment]
@commonism scroll down to the CI checks in this PR, and you'll see the status from Read The Docs. Click on the `Details` link. When it's red or in progress, it'll take you to the build log page. But once it succeeds, it leads to the preview.

File: CHANGES/1026.bugfix.rst [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+to match :rfc:`3986`.
+
+Previously empty segments would be removed from path,
+breaking use-cases such as
+
+.. code-block:: python
+
+  URL("https://web.archive.org/web/") / "https://github.com/"
+
+Now "/" and :spelling:ignore:`joinpath()` operation keep empty segments,
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:11]: Things within inline code markup aren't spellchecked.
```suggestion
Now ``"/"`` and ``joinpath()`` operation keep empty segments,
```

File: CHANGES/1026.bugfix.rst [hunk 8]
```diff
@@ -0,0 +1,20 @@
+Joining URLs with empty segments has been changed
+to match :rfc:`3986`.
+
+Previously empty segments would be removed from path,
+breaking use-cases such as
+
+.. code-block:: python
+
+  URL("https://web.archive.org/web/") / "https://github.com/"
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:9]:
```suggestion
   URL("https://web.archive.org/web/") / "https://github.com/"
```
(could also be double backticks inline, but a block is perhaps better, indeed)

File: CHANGES/1026.bugfix.rst [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+Now "/" and :spelling:ignore:`joinpath()` operation keep empty segments,
+but do not introduce new empty segments.
+e.g.
+
+.. code-block:: python
+
+  URL("https://example.org/") / ""
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:17]: Another misindent
```suggestion
   URL("https://example.org/") / ""
```

**Comment by commonism** [comment]
I've had CI runs with good coverage some minutes ago, now the bug is back.

File: tests/test_url.py (hunk 5 above)
Comment (Dreamsorcerer) [reply tests/test_url.py:864]: Yes, so it could be formatted as:
```
assert str(url / ...) == expected
assert str(url.joinpath(...)) == expected
```

Comment (webknjaz) [reply tests/test_url.py:864]: Yep.
```python
assert str(url / ...) == expected and str(url.joinpath(...)) == expected
```
would also work, without chaining.

**Comment by webknjaz** [comment]
> Coverage looks better when looking into the Actions results.

Yes, I integrated putting that into the job summary directly from whatever coverage.py collected. There's no codecov integration at that point, so it doesn't stand in the way in that specific place.

File: CHANGES/1026.bugfix.rst
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:1]: It's better now, but needs a bit more polishing: https://yarl--1026.org.readthedocs.build/en/1026/changes/#bug-fixes

File: CHANGES/1026.bugfix.rst [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+to match :rfc:`3986`.
+
+Previously empty segments would be removed from path,
+breaking use-cases such as
+
+.. code-block:: python
+
+   URL("https://web.archive.org/web/") / "https://github.com/"
+
+Now :meth:`~yarl.URL.__truediv__` and :meth:`~yarl.URL.joinpath` operation
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:11]: How about something like this?
```suggestion
Now :meth:`/ <yarl.URL.__truediv__>` and :meth:`URL.joinpath() <yarl.URL.joinpath>` operation
```

@commonism would such rendering make sense to you?

File: CHANGES/1026.bugfix.rst [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+keep empty segments, but do not introduce new empty segments.
+e.g.
+
+.. code-block:: python
+
+   URL("https://example.org/") / ""
+
+does not introduce an empty segment.
+-- by :user:`commonism` and :user:`youtux`.
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:20]: Since you're not putting this into a sentence, move it to a separate line and drop that trailing period.
```suggestion

-- by :user:`commonism` and :user:`youtux`
```

File: docs/api.rst [hunk 12]
```diff
@@ -844,6 +844,20 @@ The path is encoded if needed.
 
    .. versionadded:: 1.9
 
+.. method:: URL.__truediv__(url)
+
+   Shortcut for :meth:`URL.joinpath` with a single element and ``encoded=False``.
+
+   .. doctest::
```
Comment (webknjaz) [inline docs/api.rst:851]: Nice! I think this is the first doctest in the project. Though, we'd probably need to put a doctest sphinx invocation into CI for it to actually be checked…

File: docs/api.rst [hunk 13]
```diff
@@ -870,6 +884,8 @@ The path is encoded if needed.
          >>> base.join(URL('//python.org/page.html'))
          URL('http://python.org/page.html')
 
+
+
```
Comment (webknjaz) [inline docs/api.rst:888]: Let's drop the unrelated formatting
```suggestion
```

File: docs/api.rst [hunk 14]
```diff
@@ -844,6 +844,20 @@ The path is encoded if needed.
 
    .. versionadded:: 1.9
 
+.. method:: URL.__truediv__(url)
```
Comment (webknjaz) [inline docs/api.rst:847]: @commonism since you documented this, could you also edit the mentions of `__truediv__` in https://github.com/aio-libs/yarl/blob/master/CHANGES.rst to make them linkable like in your changelog fragment.

Additionally, add the second change note named `1026.doc.rst` mentioning that this has been documented.

File: yarl/_url.py [hunk 15]
```diff
@@ -713,28 +713,33 @@ def _validate_authority_uri_abs_path(host, path):
...
-            elif seg != ".":
-                parsed.append(seg)
+            path = path if encoded else self._PATH_QUOTER(path)
+            segments = [
+                segment for segment in reversed(path.split("/")) if segment != "."
+            ]
+            if not segments:
+                continue
+            # remove trailing empty segment for all but the last path
+            parsed += segments[1:] if not last and segments[0] == "" else segments
```
Comment (webknjaz) [inline yarl/_url.py:736]: Perhaps, splitting instructions here would also help:
```suggestion
            segment_slice_start = int(not last and segments[0] == "")
            parsed += segments[segment_slice_start:]
```

File: CHANGES/1026.bugfix.rst (hunk 11 above)
Comment (webknjaz) [reply CHANGES/1026.bugfix.rst:20]: (alternatively, you could append the byline to the very first sentence, just don't put a period before it if you decide to go that route)

File: CHANGES/1026.doc.rst [hunk 16]
```diff
@@ -0,0 +1,3 @@
+Add documentation for :meth:`/ <yarl.URL.__truediv__>`
+
+-- by :user:`commonism`
```
Comment (webknjaz) [inline CHANGES/1026.doc.rst:3]: This shouldn't be in imperative mood since it's not a commit message, nor is it a TODO list item. Also, there should be a trailing period. And since it's just one sentence, it's probably not worth separating the byline so much:
```suggestion
The pre-existing :meth:`/ <yarl.URL.__truediv__>` magic method
has been documented in the API reference -- by :user:`commonism`.
```

File: CHANGES.rst [hunk 17]
```diff
@@ -105,7 +105,7 @@ Bug fixes
 ---------
 
 - Stopped dropping trailing slashes in :py:meth:`~yarl.URL.joinpath` -- by :user:`gmacon`. (:issue:`862`, :issue:`866`)
-- Started accepting string subclasses in ``__truediv__()`` operations (``URL / segment``) -- by :user:`mjpieters`. (:issue:`871`, :issue:`884`)
+- Started accepting string subclasses in :meth:`__truediv__() <yarl.URL.__truediv__>` operations (``URL / segment``) -- by :user:`mjpieters`. (:issue:`871`, :issue:`884`)
```
Comment (webknjaz) [inline CHANGES.rst:108]:
```suggestion
- Started accepting string subclasses in :meth:`~yarl.URL.__truediv__` operations (``URL / segment``) -- by :user:`mjpieters`. (:issue:`871`, :issue:`884`)
```

File: CHANGES/1026.bugfix.rst [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+to match :rfc:`3986`.
+
+Previously empty segments would be removed from path,
+breaking use-cases such as
+
+.. code-block:: python
+
+   URL("https://web.archive.org/web/") / "https://github.com/"
+
+Now :meth:`/ <yarl.URL.__truediv__>` and :meth:`URL.joinpath() <yarl.URL.joinpath>` operation
```
Comment (webknjaz) [inline CHANGES/1026.bugfix.rst:11]: I think, this wording might be clearer:
```suggestion
Now :meth:`/ operation <yarl.URL.__truediv__>` and :meth:`URL.joinpath() <yarl.URL.joinpath>`
```

File: CHANGES/1026.doc.rst [hunk 19]
```diff
@@ -0,0 +1,2 @@
+The pre-existing :meth:`/ <yarl.URL.__truediv__>` magic method
```
Comment (webknjaz) [inline CHANGES/1026.doc.rst:1]: Just realized that it's not obvious when it's only one char `/` that is clickable:
```suggestion
The pre-existing :meth:`/ magic method <yarl.URL.__truediv__>`
```

**Review by webknjaz — APPROVED** [review]
